### ISSUE TITLE
[Feature] HTML Reporter: UI Templating/Plugins

### ISSUE THREAD

**Issue by rwoll** [microsoft/playwright#11319]
NB: 🗒️  https://github.com/microsoft/playwright/issues/11318 is related, but #11318 can be done today and is very specific/actionable, while this request is a bit longer term/extensible, but also decidedly more complicated and design-heavy.

## Request

The HTML Reports are fantastic as is. However, now that we use them and receive them daily there are some customizations we'd love to make, but we don't want to maintain an entire HTML reporter, so I'd like to propose a plugin/template system. With today's HTML Report, there's not really a place, except attachments to embed context/debug/trace information, and it's always rendered as plain text.

We'd love to add our own company specific debug tools/links/etc. in a header block[^1] as well as define a custom rendering/visualization for some test attachments. Hopefully others would like this too :)

Some additional examples:

1. Render clickable links back to the products/resources that the test created. When you're debugging a an e2e smoketest that was done on a live environment, you can quickly jump to it from the report. This way, you don't have to skim through a bunch of stdout/other attachments for an ID or URL to copy.
2. Render JSON attachments in a JSON Component Viewer. We attach big JSON blobs from our API responses, and looking at them in the report currently is less than ideal.  

(Modulo the some different styling encapsulation) here's wha the attachments section could look like with the user-authored plugins:

<img width="1033" alt="Screen Shot 2022-01-10 at 8 11 38 PM" src="https://user-images.githubusercontent.com/11915034/148880208-bd4d4a1b-67c7-4f67-bc4d-2fe44b3bff07.png">

👋  **I'm very interested in bringing this into reality in some form (see below for one quick prototype/idea[^3]), so please let me know your thoughts at both the idea and possible prioritization as it is a bigger lift and requires some design iterations on the public API.** (I'm linking to code, but I'd like to focus on whether the idea/concept/constraints rather than an implementation/solution for now.) The prototype is not the implementation nor API I would necessarily propose, but it was fun to experiment and it would be interesting to iterate on an actual public API and implementation once that's decided.

Thanks!

## Prototype

I prototyped this a bit to see what it might feel like with some initial goals of:

1. Not exposing too much API surface area (i.e. not requiring the Playwright Team to maintain a React component library). The bigger the API surface, the more to document/test/troubleshoot/etc.
2. Let users write in React since many people working with Playwright, work with React, and React has lots of docs/support.
3. Continue producing a static, single-file final artifact bundle.
4. Bolted onto the current implementation :) so it can be experimental. Don't require all the source artifacts used to produce basic report to be re-run through Webpacl. i.e. treat the user-land plugins as a 3rd-party UI lib from (the HTML Report view).
5. Auto-configured (i.e. don't make users think about how to build/bundle React bits).

I arrived at an API that looks like the following as a prototype/conversation point[^2]:

⭐  See the following for an example of what the user would write in their project: https://github.com/rwoll/playwright/tree/feature/html-report-plugins/example-report-template

tl;dr: In their project folder, they write a `index.(j|t)sx?` whose default export satisfied an interface that has optional component definitions and an optional `renderAttachment(a: Attachment) => JSX.Element` function.

If they wanted to add a JSON component viewer for JSON attachments starting from that example, they could simply do something like:

```
$ npm i react-json-view
```

and then add the following to their `renderAttachment` function:

```ts
import ReactJson from 'react-json-view';
…
const renderAttachment: Components['renderAttachment'] = (attachment: TestAttachment) => {
  …
  if (attachment.contentType === "application/json" && attachment.body) {
    return <ReactJson src={JSON.parse(attachment.body)} />
  }
  return null;
};

export default renderAttachment;
```

See the [PluginComponent](https://github.com/rwoll/playwright/blob/2a19ee183ecbda7bb513405c5fcb3d8b755c5c61/packages/html-reporter/src/PluginComponent.tsx#L40) for how the HTML Report renders the user-defined bundled components as an external/runtime lib/system. The user would know nothing about this. This component simply finds a potential plugin implementation and renders it while passing it some props, etc.
 
And see the following for how it's all tied together to produce the final bundle: https://github.com/rwoll/playwright/blob/2a19ee183ecbda7bb513405c5fcb3d8b755c5c61/packages/playwright-test/src/reporters/html.ts#L383.

In reality, depending on what constraints we wanted to impose on the architecture, I'd implement this **very differently**. Right now, the implementation bundles and treats the user-land plugins as a "third-party JS UI" library, so the base HTML Report has been changed slightly so it can call into/render the "third-party" library that gets inlined as a distinct library. (Yes, that means React is in there twice :P, but this was just for fun :P.) However, if we didn't pre-bundle the first report and bundled everything together at runtime, life is simpler, but that was less fun to try to get working :) Anyways, the implementation isn't important atm 😄 

[^1]: https://github.com/microsoft/playwright/issues/11318 would get us far for now though without all this complication!
[^2]: Some parts of it I like, some parts are very meh. Plus, perhaps there are proper templating solutions out there for this kind of thing. This was just quick prototype to get some conversations/ideas going!
[^3]: Warrants actual thinking and design hehe.

**Comment by dgozman** [comment]
Thank you for such a thorough write-up and prototyping. This is very exciting, but as you correctly say, not happening in the short term. Once we figure out the built-in features in the html report, we'll have a good understanding of the extensibility perimeter.

**Comment by ShanmukhaGajula** [comment]
This is Awesome!!

**Comment by MartinP-Autopilot** [comment]
This would be awesome.  I am working on getting our tests running in BrowserStack and I have been able to attach the BrowserStack URL for the test's session to the result in the HTML Reporter, but it is coming out as a long string of text, even if I wrap it in HTML tags - 

```
testInfo.attach('BrowserStack URL: ', {
      contentType: 'text/html',
     body: `<html><body><a href="${sessionUrl}" target="_blank">Open</a></body></html>`,  
    });
```
It is a bit onerous to copy that text to put it in the browser's address, especially when it is already in a render page.

**Comment by unlikelyzero** [comment]
This would be great for associating with percy runs

**Comment by paulovpereira** [comment]
I also need a way to customize the HTML report for our company needs, but to be honest it would be already amazing if we could just create our own class that would extend PW HTML report class just like it's explained here: https://playwright.dev/docs/test-reporters#custom-reporters

HTML reporter class needs to be exported for that to be possible.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you disagree with this closure, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.

**Comment by weilbith** [comment]
I stumbled over this issue because I somehow try to figure out a good way to "accept" failing visual regression tests. I'm using component tests in Playwright and do `toHaveScreenshot`. I use the `html` reporter that automatically opens in my browser if a test is failing and I can beautifully see the diff of the images etc. **But** there is no easy way to say as developer: this is what I intended to change. Like any regular change in behavior gets reflected in the unit tests (usually first the test, then the code).
Ofc, I could simply use `--update-snapshots`. But that is a bit too much for me and it would be too easy to accept unintended changes. Using it with `---grep="test name"` does the job, but is very cumbersome to do manually. So I would love to inject a button into the HTML report at the diff that basicially does that for me if I click it.

**Comment by krukonshedul** [comment]
I was in dire need of two things: adding some links into the report, and getting the ability to navigate the failures by error, not by failed tests. The whole Playwright Report is basically a JS that renders the HTML afterwards, so naive modifying of the HTML file couldn't address my needs. I made myself a custom reporter to organize failures by errors instead of `by failed tests`, then I wrote a quick and ugly script that allowed me to extend the report rendering logic to link to the information I've needed.

```import fs from 'fs';
import path from 'path';

const errorsReportUrl = process.argv[2];
const jobUrl = process.argv[3];

const injectedScript = `
<script>
  function applyModification() {
    const targetElement = document.querySelector('[data-testid="overall-time"]');
    if (targetElement) {
      const previousElement = targetElement.previousElementSibling;

      if (!previousElement.hasAttribute('loaded')) {
        const innerHtml = \`
          <a href="${jobUrl}">🔗 Job</a> |
          <a href="${errorsReportUrl}">Errors</a>
        \`;

        previousElement.innerHTML = innerHtml;
        previousElement.setAttribute('loaded', 'true');
      }
    }
  }
  setTimeout(() => {
    applyModification();
    setInterval(applyModification, 3000);
  }, 1000);
</script>
`;

const filePath = path.join(process.cwd(), 'playwright-report', 'index.html');
fs.readFile(filePath, 'utf-8', (err, data) => {
  if (err) {
    console.error('Error reading index.html:', err);
    return;
  }

  const modifiedHtml = data.replace('</body>', `${injectedScript}\n</body>`);

  fs.writeFile(filePath, modifiedHtml, 'utf-8', err => {
    if (err) {
      console.error('Error writing to index.html:', err);
    } else {
      console.log('index.html modified successfully.');
    }
  });
});
```

It's infinitely worse than the proper solution prototyped by the OP, but for as long as your needs don't cross the "I need to link additional information outside of test annotations" use-case, it does the job.
![image](https://github.com/user-attachments/assets/ea258955-7349-4b0c-9524-ff59aac0bfb4)

**Comment by yairEO** [comment]
I also want to be able to edit the HTML report. I want to add a "download screenshots" which will download all the "actual" screenshots (for failed tests) with their correct names, so I could simply replace my local files quickly and commit the replaces screenshot files
